{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "read_to_681_> topology.submodel read from topology file:        3\n"
     ]
    }
   ],
   "source": [
    "from modeller import *\n",
    "from modeller.scripts import complete_pdb\n",
    "\n",
    "# Initialize Modeller environment\n",
    "env = Environ()  # Use Environ instead of environ\n",
    "env.io.atom_files_directory = ['C:', r'\\Users\\Owner\\Downloads']\n",
    "\n",
    "# Read in topology and parameter libraries\n",
    "env.libs.topology.read(file='$(LIB)/top_heav.lib')\n",
    "env.libs.parameters.read(file='$(LIB)/par.lib')\n",
    "\n",
    "# Create a new model\n",
    "mdl = complete_pdb(env, '1bpv.pdb')\n",
    "\n",
    "# Display the protein structure\n",
    "mdl.write('your_protein_display.pdb')  # Save a new PDB file with the visualization information\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "from modeller import *\n",
    "from modeller.scripts import complete_pdb\n",
    "\n",
    "env = Environ()\n",
    "env.io.atom_files_directory = ['C:', r'\\Users\\Owner\\Downloads']\n",
    "\n",
    "# Simply read the model\n",
    "mdl = Model(env, file='1bpv.pdb')\n",
    "\n",
    "# Write out the model\n",
    "mdl.write(file='output.pdb')\n",
    "\n",
    "# Read the model again"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: 'sequence.fasta'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[1;32mUntitled-1.ipynb Cell 2\u001b[0m line \u001b[0;36m3\n\u001b[0;32m      <a href='vscode-notebook-cell:Untitled-1.ipynb?jupyter-notebook#W1sdW50aXRsZWQ%3D?line=0'>1</a>\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mBio\u001b[39;00m \u001b[39mimport\u001b[39;00m SeqIO\n\u001b[1;32m----> <a href='vscode-notebook-cell:Untitled-1.ipynb?jupyter-notebook#W1sdW50aXRsZWQ%3D?line=2'>3</a>\u001b[0m records \u001b[39m=\u001b[39m SeqIO\u001b[39m.\u001b[39;49mparse(\u001b[39m\"\u001b[39;49m\u001b[39msequence.fasta\u001b[39;49m\u001b[39m\"\u001b[39;49m, \u001b[39m\"\u001b[39;49m\u001b[39mfasta\u001b[39;49m\u001b[39m\"\u001b[39;49m)\n\u001b[0;32m      <a href='vscode-notebook-cell:Untitled-1.ipynb?jupyter-notebook#W1sdW50aXRsZWQ%3D?line=3'>4</a>\u001b[0m count \u001b[39m=\u001b[39m SeqIO\u001b[39m.\u001b[39mwrite(records, \u001b[39m\"\u001b[39m\u001b[39msequence.clustal\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mclustal\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[0;32m      <a href='vscode-notebook-cell:Untitled-1.ipynb?jupyter-notebook#W1sdW50aXRsZWQ%3D?line=4'>5</a>\u001b[0m \u001b[39mprint\u001b[39m(\u001b[39m\"\u001b[39m\u001b[39mConverted \u001b[39m\u001b[39m%i\u001b[39;00m\u001b[39m records\u001b[39m\u001b[39m\"\u001b[39m \u001b[39m%\u001b[39m count)\n",
      "File \u001b[1;32mc:\\Users\\Owner\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\Bio\\SeqIO\\__init__.py:605\u001b[0m, in \u001b[0;36mparse\u001b[1;34m(handle, format, alphabet)\u001b[0m\n\u001b[0;32m    603\u001b[0m iterator_generator \u001b[39m=\u001b[39m _FormatToIterator\u001b[39m.\u001b[39mget(\u001b[39mformat\u001b[39m)\n\u001b[0;32m    604\u001b[0m \u001b[39mif\u001b[39;00m iterator_generator:\n\u001b[1;32m--> 605\u001b[0m     \u001b[39mreturn\u001b[39;00m iterator_generator(handle)\n\u001b[0;32m    606\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mformat\u001b[39m \u001b[39min\u001b[39;00m AlignIO\u001b[39m.\u001b[39m_FormatToIterator:\n\u001b[0;32m    607\u001b[0m     \u001b[39m# Use Bio.AlignIO to read in the alignments\u001b[39;00m\n\u001b[0;32m    608\u001b[0m     \u001b[39mreturn\u001b[39;00m (r \u001b[39mfor\u001b[39;00m alignment \u001b[39min\u001b[39;00m AlignIO\u001b[39m.\u001b[39mparse(handle, \u001b[39mformat\u001b[39m) \u001b[39mfor\u001b[39;00m r \u001b[39min\u001b[39;00m alignment)\n",
      "File \u001b[1;32mc:\\Users\\Owner\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\Bio\\SeqIO\\FastaIO.py:223\u001b[0m, in \u001b[0;36mFastaIterator.__init__\u001b[1;34m(self, source, alphabet, title2ids)\u001b[0m\n\u001b[0;32m    204\u001b[0m     warnings\u001b[39m.\u001b[39mwarn(\n\u001b[0;32m    205\u001b[0m         \u001b[39m\"\u001b[39m\u001b[39mThe title2ids argument is deprecated. Instead, please use a \u001b[39m\u001b[39m\"\u001b[39m\n\u001b[0;32m    206\u001b[0m         \u001b[39m\"\u001b[39m\u001b[39mgenerator function to modify records returned by the parser. \u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    220\u001b[0m         BiopythonDeprecationWarning,\n\u001b[0;32m    221\u001b[0m     )\n\u001b[0;32m    222\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mtitle2ids \u001b[39m=\u001b[39m title2ids\n\u001b[1;32m--> 223\u001b[0m \u001b[39msuper\u001b[39;49m()\u001b[39m.\u001b[39;49m\u001b[39m__init__\u001b[39;49m(source, mode\u001b[39m=\u001b[39;49m\u001b[39m\"\u001b[39;49m\u001b[39mt\u001b[39;49m\u001b[39m\"\u001b[39;49m, fmt\u001b[39m=\u001b[39;49m\u001b[39m\"\u001b[39;49m\u001b[39mFasta\u001b[39;49m\u001b[39m\"\u001b[39;49m)\n",
      "File \u001b[1;32mc:\\Users\\Owner\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages\\Bio\\SeqIO\\Interfaces.py:45\u001b[0m, in \u001b[0;36mSequenceIterator.__init__\u001b[1;34m(self, source, alphabet, mode, fmt)\u001b[0m\n\u001b[0;32m     43\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mValueError\u001b[39;00m(\u001b[39m\"\u001b[39m\u001b[39mThe alphabet argument is no longer supported\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[0;32m     44\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m---> 45\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mstream \u001b[39m=\u001b[39m \u001b[39mopen\u001b[39;49m(source, \u001b[39m\"\u001b[39;49m\u001b[39mr\u001b[39;49m\u001b[39m\"\u001b[39;49m \u001b[39m+\u001b[39;49m mode)\n\u001b[0;32m     46\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mshould_close_stream \u001b[39m=\u001b[39m \u001b[39mTrue\u001b[39;00m\n\u001b[0;32m     47\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mTypeError\u001b[39;00m:  \u001b[39m# not a path, assume we received a stream\u001b[39;00m\n",
      "\u001b[1;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'sequence.fasta'"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "records = SeqIO.parse(\"sequence.fasta\", \"fasta\")\n",
    "count = SeqIO.write(records, \"sequence.clustal\", \"clustal\")\n",
    "print(\"Converted %i records\" % count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "openf___224_> Open           $(LIB)/restyp.lib\n",
      "openf___224_> Open           ${MODINSTALL10v4}/modlib/resgrp.lib\n",
      "rdresgr_266_> Number of residue groups:        2\n",
      "openf___224_> Open           ${MODINSTALL10v4}/modlib/sstruc.lib\n",
      "\n",
      "Dynamically allocated memory at   amaxlibraries [B,KiB,MiB]:       700202     683.791     0.668\n",
      "\n",
      "Dynamically allocated memory at   amaxlibraries [B,KiB,MiB]:       700730     684.307     0.668\n",
      "openf___224_> Open           ${MODINSTALL10v4}/modlib/resdih.lib\n",
      "\n",
      "Dynamically allocated memory at   amaxlibraries [B,KiB,MiB]:       749330     731.768     0.715\n",
      "rdrdih__263_> Number of dihedral angle types         :        9\n",
      "              Maximal number of dihedral angle optima:        3\n",
      "              Dihedral angle names                   :  Alph Phi Psi Omeg chi1 chi2 chi3 chi4 chi5\n",
      "openf___224_> Open           ${MODINSTALL10v4}/modlib/radii.lib\n",
      "\n",
      "Dynamically allocated memory at   amaxlibraries [B,KiB,MiB]:       762630     744.756     0.727\n",
      "openf___224_> Open           ${MODINSTALL10v4}/modlib/af_mnchdef.lib\n",
      "rdwilmo_274_> Mainchain residue conformation classes:  APBLE\n",
      "openf___224_> Open           ${MODINSTALL10v4}/modlib/mnch.lib\n",
      "rdclass_257_> Number of classes:        5\n",
      "openf___224_> Open           ${MODINSTALL10v4}/modlib/mnch1.lib\n",
      "openf___224_> Open           ${MODINSTALL10v4}/modlib/mnch2.lib\n",
      "openf___224_> Open           ${MODINSTALL10v4}/modlib/mnch3.lib\n",
      "openf___224_> Open           ${MODINSTALL10v4}/modlib/xs4.mat\n",
      "rdrrwgh_268_> Number of residue types:       21\n",
      "openf___224_> Open           pdb_95.pir\n",
      "\n",
      "Dynamically allocated memory at amaxsequence_db [B,KiB,MiB]:       807252     788.332     0.770\n",
      "\n",
      "Dynamically allocated memory at amaxsequence_db [B,KiB,MiB]:      1394386    1361.705     1.330\n",
      "\n",
      "Dynamically allocated memory at   amaxalignment [B,KiB,MiB]:      1396628    1363.895     1.332\n",
      "\n",
      "Dynamically allocated memory at   amaxalignment [B,KiB,MiB]:      1398078    1365.311     1.333\n",
      "\n",
      "Dynamically allocated memory at   amaxalignment [B,KiB,MiB]:      1400978    1368.143     1.336\n",
      "\n",
      "Dynamically allocated memory at amaxsequence_db [B,KiB,MiB]:      3616777    3532.009     3.449\n",
      "\n",
      "Dynamically allocated memory at   amaxalignment [B,KiB,MiB]:      3622577    3537.673     3.455\n",
      "\n",
      "Dynamically allocated memory at   amaxalignment [B,KiB,MiB]:      3634177    3549.001     3.466\n",
      "\n",
      "Dynamically allocated memory at   amaxalignment [B,KiB,MiB]:      3657377    3571.657     3.488\n",
      "\n",
      "Dynamically allocated memory at amaxsequence_db [B,KiB,MiB]:      4765277    4653.591     4.545\n",
      "\n",
      "SEQ_DATABASE_FILE         : pdb_95.pir\n",
      "SEQ_DATABASE_FORMAT       : PIR\n",
      "CHAINS_LIST               : ALL\n",
      "CLEAN_SEQUENCES           :            T\n",
      "MINMAX_DB_SEQ_LEN         :           30        4000\n",
      "Number of sequences       :        11079\n",
      "Number of residues        :      2436254\n",
      "Length of longest sequence:         1491\n",
      "\n",
      "\n",
      "SEQ_DATABASE_FILE         : pdb_95.bin\n",
      "SEQ_DATABASE_FORMAT       : BINARY\n",
      "SEARCH_CHAINS_LIST        : ALL\n",
      "Number of sequences       :        11079\n",
      "Number of residues        :      2436254\n",
      "Length of longest sequence:         1491\n",
      "\n",
      "\n",
      "SEQ_DATABASE_FILE         : pdb_95.bin\n",
      "SEQ_DATABASE_FORMAT       : BINARY\n",
      "CHAINS_LIST               : ALL\n",
      "CLEAN_SEQUENCES           :            T\n",
      "MINMAX_DB_SEQ_LEN         :            0      999999\n",
      "Number of sequences       :        11079\n",
      "Number of residues        :      2436254\n",
      "Length of longest sequence:         1491\n",
      "\n",
      "openf___224_> Open           TvLDH.ali\n",
      "\n",
      "Dynamically allocated memory at   amaxalignment [B,KiB,MiB]:       809572     790.598     0.772\n",
      "\n",
      "Dynamically allocated memory at   amaxalignment [B,KiB,MiB]:       811022     792.014     0.773\n",
      "\n",
      "Dynamically allocated memory at   amaxalignment [B,KiB,MiB]:       813922     794.846     0.776\n",
      "\n",
      "Dynamically allocated memory at   amaxalignment [B,KiB,MiB]:       819722     800.510     0.782\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:       821058     801.814     0.783\n",
      "\n",
      "Read the alignment from file       : TvLDH.ali\n",
      "\n",
      "Total number of alignment positions:   335\n",
      "\n",
      "  #  Code        #_Res #_Segm PDB_code    Name\n",
      "-------------------------------------------------------------------------------\n",
      "  1      TvLDH     335      1       TvLDH\n",
      "\n",
      "Dynamically allocated memory at     amaxprofile [B,KiB,MiB]:       822731     803.448     0.785\n",
      "openf___224_> Open           ${LIB}/blosum62.sim.mat\n",
      "rdrrwgh_268_> Number of residue types:       21\n",
      "\n",
      "Dynamically allocated memory at amaxsequence_db [B,KiB,MiB]:      1099201    1073.438     1.048\n",
      "profile_iteration_> processing sequence:       1    335      1     0.0000000     0.0000000     0.0000000     1\n",
      "\n",
      "Dynamically allocated memory at amaxsequence_db [B,KiB,MiB]:      1108498    1082.518     1.057\n",
      "profile_iteration_> processing sequence:    1109    335      1     0.3906250     0.0003522     0.0000016  1109\n",
      "\n",
      "Dynamically allocated memory at amaxsequence_db [B,KiB,MiB]:      1116286    1090.123     1.065\n",
      "profile_iteration_> processing sequence:    2217    335      1     0.7968750     0.0003594     0.0000016  2217\n",
      "profile_iteration_> processing sequence:    3325    335      1     1.2031250     0.0003618     0.0000016  2764\n",
      "profile_iteration_> processing sequence:    4433    335      1     1.6093750     0.0003630     0.0000016  2754\n",
      "profile_iteration_> processing sequence:    5541    335      1     1.9843750     0.0003581     0.0000016  2792\n",
      "profile_iteration_> processing sequence:    6649    335      1     2.3750000     0.0003572     0.0000016  2800\n",
      "profile_iteration_> processing sequence:    7757    335      1     2.7656250     0.0003565     0.0000016  2805\n",
      "profile_iteration_> processing sequence:    8865    335      1     3.1406250     0.0003543     0.0000016  2823\n",
      "profile_iteration_> processing sequence:    9973    335      1     3.5000000     0.0003509     0.0000016  2849\n",
      " AVERAGE RESIDUAL STANDARD ERROR:    27.32606\n",
      " AVERAGE RESIDUAL STANDARD DEVIATION:    243.2711\n",
      "profile_660_> Quality of the fit between observed/expected score distributions.\n",
      "              [Iteration, Chi2, nbins, KS-Stat, nscores ] =        1,        0.0584,      121,        0.0130,    11079\n",
      "\n",
      "HISTOGRAM OF Z-SCORES:\n",
      "\n",
      "Z:         1  -3.95000   0.00000   0.00000\n",
      "Z:         1  -3.85000   0.00000   0.00000\n",
      "Z:         1  -3.75000   0.00000   0.00000\n",
      "Z:         1  -3.65000   0.00000   0.00000\n",
      "Z:         1  -3.55000   0.00000   0.00000\n",
      "Z:         1  -3.45000   0.00000   0.00000\n",
      "Z:         1  -3.35000   0.00000   0.00000\n",
      "Z:         1  -3.25000   0.00000   0.00000\n",
      "Z:         1  -3.15000   0.00000   0.00000\n",
      "Z:         1  -3.05000   0.00000   0.00000\n",
      "Z:         1  -2.95000   0.00000   0.00000\n",
      "Z:         1  -2.85000   0.00000   0.00000\n",
      "Z:         1  -2.75000   0.00000   0.00000\n",
      "Z:         1  -2.65000   0.00000   0.00000\n",
      "Z:         1  -2.55000   0.00000   0.00001\n",
      "Z:         1  -2.45000   0.00090   0.00004\n",
      "Z:         1  -2.35000   0.00181   0.00016\n",
      "Z:         1  -2.25000   0.00000   0.00055\n",
      "Z:         1  -2.15000   0.00271   0.00163\n",
      "Z:         1  -2.05000   0.00722   0.00416\n",
      "Z:         1  -1.95000   0.01354   0.00933\n",
      "Z:         1  -1.85000   0.02527   0.01872\n",
      "Z:         1  -1.75000   0.05235   0.03400\n",
      "Z:         1  -1.65000   0.04603   0.05658\n",
      "Z:         1  -1.55000   0.09026   0.08721\n",
      "Z:         1  -1.45000   0.12727   0.12565\n",
      "Z:         1  -1.35000   0.16698   0.17059\n",
      "Z:         1  -1.25000   0.20218   0.21981\n",
      "Z:         1  -1.15000   0.26627   0.27052\n",
      "Z:         1  -1.05000   0.30057   0.31973\n",
      "Z:         1  -0.95000   0.36917   0.36469\n",
      "Z:         1  -0.85000   0.35111   0.40316\n",
      "Z:         1  -0.75000   0.46484   0.43360\n",
      "Z:         1  -0.65000   0.41249   0.45519\n",
      "Z:         1  -0.55000   0.47206   0.46779\n",
      "Z:         1  -0.45000   0.45853   0.47182\n",
      "Z:         1  -0.35000   0.45672   0.46811\n",
      "Z:         1  -0.25000   0.44769   0.45775\n",
      "Z:         1  -0.15000   0.50817   0.44196\n",
      "Z:         1  -0.05000   0.41430   0.42194\n",
      "Z:         1   0.05000   0.39173   0.39889\n",
      "Z:         1   0.15000   0.37910   0.37383\n",
      "Z:         1   0.25000   0.35834   0.34769\n",
      "Z:         1   0.35000   0.29696   0.32121\n",
      "Z:         1   0.45000   0.28613   0.29500\n",
      "Z:         1   0.55000   0.28703   0.26953\n",
      "Z:         1   0.65000   0.21933   0.24513\n",
      "Z:         1   0.75000   0.22836   0.22205\n",
      "Z:         1   0.85000   0.21211   0.20043\n",
      "Z:         1   0.95000   0.18594   0.18036\n",
      "Z:         1   1.05000   0.18233   0.16185\n",
      "Z:         1   1.15000   0.14622   0.14489\n",
      "Z:         1   1.25000   0.13900   0.12944\n",
      "Z:         1   1.35000   0.10561   0.11542\n",
      "Z:         1   1.45000   0.08936   0.10275\n",
      "Z:         1   1.55000   0.10199   0.09133\n",
      "Z:         1   1.65000   0.07762   0.08109\n",
      "Z:         1   1.75000   0.08485   0.07191\n",
      "Z:         1   1.85000   0.07672   0.06371\n",
      "Z:         1   1.95000   0.06409   0.05639\n",
      "Z:         1   2.05000   0.03340   0.04988\n",
      "Z:         1   2.15000   0.04152   0.04409\n",
      "Z:         1   2.25000   0.04513   0.03895\n",
      "Z:         1   2.35000   0.02708   0.03439\n",
      "Z:         1   2.45000   0.03069   0.03035\n",
      "Z:         1   2.55000   0.02257   0.02678\n",
      "Z:         1   2.65000   0.01895   0.02361\n",
      "Z:         1   2.75000   0.01534   0.02082\n",
      "Z:         1   2.85000   0.01444   0.01835\n",
      "Z:         1   2.95000   0.01444   0.01617\n",
      "Z:         1   3.05000   0.01264   0.01424\n",
      "Z:         1   3.15000   0.01534   0.01255\n",
      "Z:         1   3.25000   0.01354   0.01105\n",
      "Z:         1   3.35000   0.01173   0.00973\n",
      "Z:         1   3.45000   0.00722   0.00857\n",
      "Z:         1   3.55000   0.00993   0.00754\n",
      "Z:         1   3.65000   0.00722   0.00664\n",
      "Z:         1   3.75000   0.00361   0.00584\n",
      "Z:         1   3.85000   0.00361   0.00514\n",
      "Z:         1   3.95000   0.00271   0.00453\n",
      "Z:         1   4.05000   0.00993   0.00398\n",
      "Z:         1   4.15000   0.00542   0.00350\n",
      "Z:         1   4.25000   0.00271   0.00308\n",
      "Z:         1   4.35000   0.00451   0.00271\n",
      "Z:         1   4.45000   0.00181   0.00239\n",
      "Z:         1   4.55000   0.00361   0.00210\n",
      "Z:         1   4.65000   0.00271   0.00185\n",
      "Z:         1   4.75000   0.00090   0.00163\n",
      "Z:         1   4.85000   0.00000   0.00143\n",
      "Z:         1   4.95000   0.00361   0.00126\n",
      "Z:         1   5.05000   0.00090   0.00111\n",
      "Z:         1   5.15000   0.00090   0.00097\n",
      "Z:         1   5.25000   0.00271   0.00086\n",
      "Z:         1   5.35000   0.00181   0.00075\n",
      "Z:         1   5.45000   0.00000   0.00066\n",
      "Z:         1   5.55000   0.00271   0.00058\n",
      "Z:         1   5.65000   0.00090   0.00051\n",
      "Z:         1   5.75000   0.00181   0.00045\n",
      "Z:         1   5.85000   0.00090   0.00040\n",
      "Z:         1   5.95000   0.00000   0.00035\n",
      "Z:         1   6.05000   0.00090   0.00031\n",
      "Z:         1   6.15000   0.00090   0.00027\n",
      "Z:         1   6.25000   0.00000   0.00024\n",
      "Z:         1   6.35000   0.00000   0.00021\n",
      "Z:         1   6.45000   0.00000   0.00018\n",
      "Z:         1   6.55000   0.00000   0.00016\n",
      "Z:         1   6.65000   0.00000   0.00014\n",
      "Z:         1   6.75000   0.00000   0.00013\n",
      "Z:         1   6.85000   0.00000   0.00011\n",
      "Z:         1   6.95000   0.00090   0.00010\n",
      "Z:         1   7.05000   0.00000   0.00009\n",
      "Z:         1   7.15000   0.00000   0.00007\n",
      "Z:         1   7.25000   0.00000   0.00007\n",
      "Z:         1   7.35000   0.00000   0.00006\n",
      "Z:         1   7.45000   0.00000   0.00005\n",
      "Z:         1   7.55000   0.00000   0.00004\n",
      "Z:         1   7.65000   0.00000   0.00004\n",
      "Z:         1   7.75000   0.00000   0.00003\n",
      "Z:         1   7.85000   0.00000   0.00003\n",
      "Z:         1   7.95000   0.00000   0.00003\n",
      "Z:         1   8.05000   0.00000   0.00002\n",
      "\n",
      "HITS FOUND IN ITERATION:     1\n",
      "\n",
      "\n",
      "Dynamically allocated memory at     amaxprofile [B,KiB,MiB]:      1401286    1368.443     1.336\n",
      "> 1a5z                       1     190    6900     312     335   28.14    0.85E-08       2   164    75   242    63   229\n",
      "> 1b8pA                      1     951   29400     327     335   42.50     0.0           3   316     7   331     6   325\n",
      "> 1bdmA                      1    1058   30700     318     335   44.52     0.0           4   309     1   325     1   310\n",
      "> 1t2dA                      1    1717    5200     315     335   24.70    0.67E-04       5   238     5   256     4   250\n",
      "> 1civA                      1    1784   23950     374     335   34.66     0.0           6   325     6   334    33   358\n",
      "> 2cmd                       1    1858    5900     312     335   27.24    0.17E-05       7   289     7   320     3   303\n",
      "> 1o6zA                      1    2060    5800     303     335   26.32    0.27E-05       8   278     7   320     3   287\n",
      "> 1ur5A                      1    3954    4500     299     335   30.67    0.25E-02       9   158    13   191     9   171\n",
      "> 1guzA                      1    3955    7100     305     335   25.27    0.29E-08      10   265    13   301     8   280\n",
      "> 1gv0A                      1    3962    5350     301     335   25.53    0.29E-04      11   274    13   323     8   289\n",
      "> 1hyeA                      1    4405    6800     307     335   29.28    0.14E-07      12   173     7   191     3   183\n",
      "> 1i0zA                      1    4433    5650     332     335   24.64    0.67E-05      13   207    85   300    94   304\n",
      "> 1i10A                      1    4435    5600     331     335   26.21    0.87E-05      14   196    85   295    93   298\n",
      "> 1ldnA                      1    5824    5000     316     335   25.76    0.19E-03      15   214    78   298    73   301\n",
      "> 6ldh                       1    5831    4600     329     335   23.08    0.17E-02      16   244    47   301    56   302\n",
      "> 2ldx                       1    5833    5400     331     335   25.83    0.25E-04      17   227    66   306    67   306\n",
      "> 5ldh                       1    5835    5800     333     335   25.59    0.30E-05      18   207    85   300    94   304\n",
      "> 9ldtA                      1    5836    6000     331     335   25.94    0.11E-05      19   207    85   301    93   304\n",
      "> 1llc                       1    5896    5000     321     335   25.82    0.20E-03      20   164    64   239    53   234\n",
      "> 1lldA                      1    5897    6650     313     335   30.67    0.32E-07      21   216    13   242     9   233\n",
      "> 5mdhA                      1    6084   32900     333     335   44.41     0.0          22   328     2   332     1   331\n",
      "> 7mdhA                      1    6085   22600     351     335   33.74     0.0          23   325     6   334    14   339\n",
      "> 1mldA                      1    6188    6100     313     335   25.93    0.58E-06      24   183     5   198     1   189\n",
      "> 1oc4A                      1    6736    5450     315     335   27.87    0.18E-04      25   174     5   191     4   186\n",
      "> 1ojuA                      1    6799    5700     294     335   27.98    0.44E-05      26   218    78   320    68   285\n",
      "> 1pzgA                      1    7253    6350     327     335   30.00    0.16E-06      27   114    74   191    71   190\n",
      "> 1smkA                      1    7850    8750     313     335   33.85     0.0          28   188     7   202     4   198\n",
      "> 1sovA                      1    7980    4700     316     335   26.59    0.94E-03      29   160    81   256    76   248\n",
      "> 1y6jA                      1   10912    5750     289     335   32.73    0.33E-05      30   109    77   191    58   167\n",
      "\n",
      "profile_iteration__> Weight Matrix           [sec]:     0.000000\n",
      "profile_iteration__> Score Collection        [sec]:     3.890625\n",
      "profile_iteration__> Score Collection/Seq    [sec]:     0.000351\n",
      "profile_iteration__> DP Steps per sec    [ali/sec]:  2859.096680\n",
      "profile_iteration__> Statistics              [sec]:     0.015625\n",
      "profile_iteration__> Add to profile          [sec]:     0.031250\n",
      "profile_iteration__> Overall performance    [/seq]:     0.000355\n",
      "profile_iteration__> Overall performance [seq/sec]:  2813.714355\n",
      "openf___224_> Open           build_profile.prf\n",
      "\n",
      "Dynamically allocated memory at   amaxalignment [B,KiB,MiB]:      1222305    1193.657     1.166\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:      1223641    1194.962     1.167\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:      1224293    1195.599     1.168\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:      1225553    1196.829     1.169\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:      1226785    1198.032     1.170\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:      1227733    1198.958     1.171\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:      1229029    1200.224     1.172\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:      1230181    1201.349     1.173\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:      1231289    1202.431     1.174\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:      1231917    1203.044     1.175\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:      1232973    1204.075     1.176\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:      1234065    1205.142     1.177\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:      1234753    1205.813     1.178\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:      1235577    1206.618     1.178\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:      1236357    1207.380     1.179\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:      1237209    1208.212     1.180\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:      1238181    1209.161     1.181\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:      1239085    1210.044     1.182\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:      1239909    1210.849     1.182\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:      1240733    1211.653     1.183\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:      1241385    1212.290     1.184\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:      1242245    1213.130     1.185\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:      1243553    1214.407     1.186\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:      1244849    1215.673     1.187\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:      1245577    1216.384     1.188\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:      1246269    1217.060     1.189\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:      1247137    1217.907     1.189\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:      1247589    1218.349     1.190\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:      1248337    1219.079     1.191\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:      1248973    1219.700     1.191\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:      1249405    1220.122     1.192\n",
      "openf___224_> Open           build_profile.ali\n"
     ]
    }
   ],
   "source": [
    "from modeller import *\n",
    "\n",
    "log.verbose()\n",
    "env = Environ()\n",
    "\n",
    "#-- Prepare the input files\n",
    "\n",
    "#-- Read in the sequence database\n",
    "sdb = SequenceDB(env)\n",
    "sdb.read(seq_database_file='pdb_95.pir', seq_database_format='PIR',\n",
    "         chains_list='ALL', minmax_db_seq_len=(30, 4000), clean_sequences=True)\n",
    "\n",
    "#-- Write the sequence database in binary form\n",
    "sdb.write(seq_database_file='pdb_95.bin', seq_database_format='BINARY',\n",
    "          chains_list='ALL')\n",
    "\n",
    "#-- Now, read in the binary database\n",
    "sdb.read(seq_database_file='pdb_95.bin', seq_database_format='BINARY',\n",
    "         chains_list='ALL')\n",
    "\n",
    "#-- Read in the target sequence/alignment\n",
    "aln = Alignment(env)\n",
    "aln.append(file='TvLDH.ali', alignment_format='PIR', align_codes='ALL')\n",
    "\n",
    "#-- Convert the input sequence/alignment into\n",
    "#   profile format\n",
    "prf = aln.to_profile()\n",
    "\n",
    "#-- Scan sequence database to pick up homologous sequences\n",
    "prf.build(sdb, matrix_offset=-450, rr_file='${LIB}/blosum62.sim.mat',\n",
    "          gap_penalties_1d=(-500, -50), n_prof_iterations=1,\n",
    "          check_profile=False, max_aln_evalue=0.01)\n",
    "\n",
    "#-- Write out the profile in text format\n",
    "prf.write(file='build_profile.prf', profile_format='TEXT')\n",
    "\n",
    "#-- Convert the profile back to alignment format\n",
    "aln = prf.to_alignment()\n",
    "\n",
    "#-- Write out the alignment file\n",
    "aln.write(file='build_profile.ali', alignment_format='PIR')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "                         MODELLER 10.4, 2022/10/28, r12463\n",
      "\n",
      "     PROTEIN STRUCTURE MODELLING BY SATISFACTION OF SPATIAL RESTRAINTS\n",
      "\n",
      "\n",
      "                     Copyright(c) 1989-2022 Andrej Sali\n",
      "                            All Rights Reserved\n",
      "\n",
      "                             Written by A. Sali\n",
      "                               with help from\n",
      "              B. Webb, M.S. Madhusudhan, M-Y. Shen, G.Q. Dong,\n",
      "          M.A. Marti-Renom, N. Eswar, F. Alber, M. Topf, B. Oliva,\n",
      "             A. Fiser, R. Sanchez, B. Yerkovich, A. Badretdinov,\n",
      "                     F. Melo, J.P. Overington, E. Feyfant\n",
      "                 University of California, San Francisco, USA\n",
      "                    Rockefeller University, New York, USA\n",
      "                      Harvard University, Cambridge, USA\n",
      "                   Imperial Cancer Research Fund, London, UK\n",
      "              Birkbeck College, University of London, London, UK\n",
      "\n",
      "\n",
      "Kind, OS, HostName, Kernel, Processor: 4, WinNT 10.0 build 19045, DESKTOP-TC0O6H7, SMP, unknown\n",
      "Date and time of compilation         : 2022/10/28 21:14:25\n",
      "MODELLER executable type             : x86_64-w64\n",
      "Job starting time (YY/MM/DD HH:MM:SS): 2023/09/28 13:59:51\n",
      "\n",
      "pdbnam_____E> Filename for PDB code not found: 1b8p\n",
      "              Directories: \n",
      "              Extensions : ;.atm;.pdb;.ent;.cif;.crd\n",
      "              (Also tried prepending 'pdb', looking for .Z, .gz, .bz2, .7z, .xz,\n",
      "              and trying PDB-style subdirectories - e.g. ab for pdb1abc.ent)\n",
      "\n"
     ]
    },
    {
     "ename": "OSError",
     "evalue": "pdbnam_____E> Filename for PDB code not found: 1b8p\n              Directories: \n              Extensions : ;.atm;.pdb;.ent;.cif;.crd\n              (Also tried prepending 'pdb', looking for .Z, .gz, .bz2, .7z, .xz,\n              and trying PDB-style subdirectories - e.g. ab for pdb1abc.ent)\n",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mOSError\u001b[0m                                   Traceback (most recent call last)",
      "\u001b[1;32mc:\\Users\\Owner\\transcripomics\\modeler_test.ipynb Cell 5\u001b[0m line \u001b[0;36m7\n\u001b[0;32m      <a href='vscode-notebook-cell:/c%3A/Users/Owner/transcripomics/modeler_test.ipynb#W4sZmlsZQ%3D%3D?line=3'>4</a>\u001b[0m aln \u001b[39m=\u001b[39m Alignment(env)\n\u001b[0;32m      <a href='vscode-notebook-cell:/c%3A/Users/Owner/transcripomics/modeler_test.ipynb#W4sZmlsZQ%3D%3D?line=4'>5</a>\u001b[0m \u001b[39mfor\u001b[39;00m (pdb, chain) \u001b[39min\u001b[39;00m ((\u001b[39m'\u001b[39m\u001b[39m1b8p\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mA\u001b[39m\u001b[39m'\u001b[39m), (\u001b[39m'\u001b[39m\u001b[39m1bdm\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mA\u001b[39m\u001b[39m'\u001b[39m), (\u001b[39m'\u001b[39m\u001b[39m1civ\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mA\u001b[39m\u001b[39m'\u001b[39m),\n\u001b[0;32m      <a href='vscode-notebook-cell:/c%3A/Users/Owner/transcripomics/modeler_test.ipynb#W4sZmlsZQ%3D%3D?line=5'>6</a>\u001b[0m                      (\u001b[39m'\u001b[39m\u001b[39m5mdh\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mA\u001b[39m\u001b[39m'\u001b[39m), (\u001b[39m'\u001b[39m\u001b[39m7mdh\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mA\u001b[39m\u001b[39m'\u001b[39m), (\u001b[39m'\u001b[39m\u001b[39m1smk\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mA\u001b[39m\u001b[39m'\u001b[39m)):\n\u001b[1;32m----> <a href='vscode-notebook-cell:/c%3A/Users/Owner/transcripomics/modeler_test.ipynb#W4sZmlsZQ%3D%3D?line=6'>7</a>\u001b[0m     m \u001b[39m=\u001b[39m Model(env, file\u001b[39m=\u001b[39;49mpdb, model_segment\u001b[39m=\u001b[39;49m(\u001b[39m'\u001b[39;49m\u001b[39mFIRST:\u001b[39;49m\u001b[39m'\u001b[39;49m\u001b[39m+\u001b[39;49mchain, \u001b[39m'\u001b[39;49m\u001b[39mLAST:\u001b[39;49m\u001b[39m'\u001b[39;49m\u001b[39m+\u001b[39;49mchain))\n\u001b[0;32m      <a href='vscode-notebook-cell:/c%3A/Users/Owner/transcripomics/modeler_test.ipynb#W4sZmlsZQ%3D%3D?line=7'>8</a>\u001b[0m     aln\u001b[39m.\u001b[39mappend_model(m, atom_files\u001b[39m=\u001b[39mpdb, align_codes\u001b[39m=\u001b[39mpdb\u001b[39m+\u001b[39mchain)\n\u001b[0;32m      <a href='vscode-notebook-cell:/c%3A/Users/Owner/transcripomics/modeler_test.ipynb#W4sZmlsZQ%3D%3D?line=8'>9</a>\u001b[0m aln\u001b[39m.\u001b[39mmalign()\n",
      "File \u001b[1;32mC:\\Program Files\\Modeller10.4\\modlib\\modeller\\model.py:101\u001b[0m, in \u001b[0;36mModel.__init__\u001b[1;34m(self, env, **vars)\u001b[0m\n\u001b[0;32m     99\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mgroup_restraints \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39menv\u001b[39m.\u001b[39mgroup_restraints\n\u001b[0;32m    100\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mlen\u001b[39m(\u001b[39mvars\u001b[39m) \u001b[39m>\u001b[39m \u001b[39m0\u001b[39m:\n\u001b[1;32m--> 101\u001b[0m     \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mread(\u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49m\u001b[39mvars\u001b[39;49m)\n",
      "File \u001b[1;32mC:\\Program Files\\Modeller10.4\\modlib\\modeller\\model.py:148\u001b[0m, in \u001b[0;36mModel.read\u001b[1;34m(self, file, model_format, model_segment, io, keep_disulfides)\u001b[0m\n\u001b[0;32m    146\u001b[0m     io \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39menv\u001b[39m.\u001b[39mio\n\u001b[0;32m    147\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39misinstance\u001b[39m(file, \u001b[39mstr\u001b[39m):\n\u001b[1;32m--> 148\u001b[0m     file \u001b[39m=\u001b[39m _modeller\u001b[39m.\u001b[39;49mmod_pdb_filename_get(file,\n\u001b[0;32m    149\u001b[0m                                           io\u001b[39m.\u001b[39;49matom_files_directory)\n\u001b[0;32m    150\u001b[0m fh \u001b[39m=\u001b[39m modfile\u001b[39m.\u001b[39m_get_filehandle(file, \u001b[39m'\u001b[39m\u001b[39mr\u001b[39m\u001b[39m'\u001b[39m)\n\u001b[0;32m    151\u001b[0m \u001b[39mreturn\u001b[39;00m _modeller\u001b[39m.\u001b[39mmod_model_read2(\u001b[39mself\u001b[39m\u001b[39m.\u001b[39mmodpt, io\u001b[39m.\u001b[39mmodpt,\n\u001b[0;32m    152\u001b[0m                                  \u001b[39mself\u001b[39m\u001b[39m.\u001b[39menv\u001b[39m.\u001b[39mlibs\u001b[39m.\u001b[39mmodpt, fh\u001b[39m.\u001b[39mfile_pointer,\n\u001b[0;32m    153\u001b[0m                                  model_format, model_segment,\n\u001b[0;32m    154\u001b[0m                                  keep_disulfides)\n",
      "\u001b[1;31mOSError\u001b[0m: pdbnam_____E> Filename for PDB code not found: 1b8p\n              Directories: \n              Extensions : ;.atm;.pdb;.ent;.cif;.crd\n              (Also tried prepending 'pdb', looking for .Z, .gz, .bz2, .7z, .xz,\n              and trying PDB-style subdirectories - e.g. ab for pdb1abc.ent)\n"
     ]
    }
   ],
   "source": [
    "from modeller import *\n",
    "\n",
    "env = Environ()\n",
    "aln = Alignment(env)\n",
    "for (pdb, chain) in (('1b8p', 'A'), ('1bdm', 'A'), ('1civ', 'A'),\n",
    "                     ('5mdh', 'A'), ('7mdh', 'A'), ('1smk', 'A')):\n",
    "    m = Model(env, file=pdb, model_segment=('FIRST:'+chain, 'LAST:'+chain))\n",
    "    aln.append_model(m, atom_files=pdb, align_codes=pdb+chain)\n",
    "aln.malign()\n",
    "aln.malign3d()\n",
    "aln.compare_structures()\n",
    "aln.id_table(matrix_file='family.mat')\n",
    "env.dendrogram(matrix_file='family.mat', cluster_cut=-1.0)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.2"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
